src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -5718,14 +5718,14 @@ int RGWRados::Object::Delete::delete_obj(optional_yield y, const DoutPrefixProvi
     }
   }
 
-  if (!state->exists) {
-    target->invalidate_state();
-    return -ENOENT;
-  }
+  bool need_invalidate = !state->exists;
+  bool exists = state->exists;
 
-  r = target->prepare_atomic_modification(dpp, op, false, NULL, NULL, NULL, true, false, y);
-  if (r < 0)
-    return r;
+  if (exists) {
+    r = target->prepare_atomic_modification(dpp, op, false, NULL, NULL, NULL, true, false, y);
```

Comment:
Double check my analysis here, but it looks to me like the `complete_atomic_modification` method is cleaning up the tail of an object and it seems that it's possible for the tail to not be cleaned up (if the process is aborted or killed or the request thread is terminated early in some other way). I was thinking that the subsequent attempts to delete the object should again attempt to delete any tail objects, in case some were left behind. It's a good point though, and a scenario that we should probably test to make sure there are no surprises.